addons/payment/models/account_journal.py

```diff
@@ -70,6 +43,24 @@ def _compute_available_payment_method_ids(self):
 
                 journal.available_payment_method_ids = [Command.unlink(payment_method) for payment_method in to_remove]
 
+    def write(self, vals):
+        """ Ensure that the user does not remove an account.payment.method.line that is linked to an
+        acquirer in the test or enabled state.
+        """
+        pre_inbound_payment_method_line_ids = self.inbound_payment_method_line_ids
+        res = super().write(vals)
+        if 'inbound_payment_method_line_ids' in vals:
+            removed_apmls = pre_inbound_payment_method_line_ids - self.inbound_payment_method_line_ids
+            active_providers = self.env['payment.acquirer'].search([
+                ('provider', 'in', removed_apmls.payment_method_id.mapped('code')),
+                ('state', 'in', ['enabled', 'test'])
+            ])
+            if active_providers:
+                raise UserError(
+                    _("You can't delete a payment method that is linked to a provider in the enabled or test state.\n"
+                      "Linked providers(s): %s", ', '.join(a.display_name for a in active_providers)))
```

Comment:
Nitpicking: while you are moving these lines, may I bother you to reformat them for better readability? We pay more and more attention to clean code and the commit will be forward-ported, so better address that now :)
```suggestion
                raise UserError(_(
                    "You can't delete a payment method that is linked to a provider in the enabled"
                    " or test state.\n""Linked providers(s): %s",
                    ', '.join(a.display_name for a in active_providers),
                ))
```

Reply:
Btw indirectly related to this PR, in the first place I replaced the constraint here with a check in `unlink()`, that raised the aforementioned error but Pylint whined on the raise in `unlink` (see [screenshot](https://cdn.discordapp.com/attachments/1067446960245710849/1126055552074657792/image.png))
Any clue why that wouldn't be possible ?

Reply:
> Have we decided on a more precise definition for formatting ?
Imo both codes here are clean, just two ways :)

No, there is no definitive guideline for this except the recommendation to follow the code style of the surrounding code/file/module. I *personally* find it easier to read when function arguments are not crammed together and closing parenthesis align well with opening parenthesis, but one could have the opposite point of view.

Because you cannot raise when uninstalling modules, a process that often calls `unlink`. This made me realize that we developed a decorator for this case. See https://github.com/odoo/odoo/pull/127350#discussion_r1255494080